File: lang/python/src/crucible/agentic.py
import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml

_PACKAGE_ROOT = Path(__file__).parent
_ROLES_DIR = _PACKAGE_ROOT.parent.parent / "config" / "agentic" / "roles"

_ROLE_SLUG_RE = re.compile(r"^[a-z0-9][a-z0-9_-]*$")


@dataclass
class RoleMindset:
    focus: list[str] = field(default_factory=list)
    principles: list[str] = field(default_factory=list)


@dataclass
class RoleEscalation:
    target: str = ""
    when: str = ""


@dataclass
class RoleExample:
    type: str = ""
    title: str = ""
    content: str = ""


@dataclass
class RolePrompt:
    """Typed representation of a role definition from the agentic role catalog."""

    slug: str = ""
    name: str = ""
    description: str = ""
    version: str = ""
    status: str = ""
    scope: list[str] = field(default_factory=list)
    responsibilities: list[str] = field(default_factory=list)
    escalates_to: list[RoleEscalation] = field(default_factory=list)
    does_not: list[str] = field(default_factory=list)
    author: str | None = None
    category: str | None = None
    extends: str | None = None
    domains: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    context: str | None = None
    mindset: RoleMindset | None = None
    examples: list[RoleExample] = field(default_factory=list)
    checklists: dict[str, list[str]] = field(default_factory=dict)


def _parse_role(data: dict) -> RolePrompt:
    mindset = None
    if "mindset" in data and data["mindset"]:
        m = data["mindset"]
        mindset = RoleMindset(
            focus=m.get("focus") or [],
            principles=m.get("principles") or [],
        )

    escalates_to = [
        RoleEscalation(target=e.get("target", ""), when=e.get("when", "")) for e in (data.get("escalates_to") or [])
    ]

    examples = [
        RoleExample(
            type=e.get("type", ""),
            title=e.get("title", ""),
            content=e.get("content", ""),
        )
        for e in (data.get("examples") or [])
    ]

    return RolePrompt(
        slug=data.get("slug", ""),
        name=data.get("name", ""),
        description=data.get("description", ""),
        version=data.get("version", ""),
        status=data.get("status", ""),
        author=data.get("author"),
        category=data.get("category"),
        extends=data.get("extends"),
        domains=data.get("domains") or [],
        tags=data.get("tags") or [],
        context=data.get("context"),
        scope=data.get("scope") or [],
        mindset=mindset,
        responsibilities=data.get("responsibilities") or [],
        escalates_to=escalates_to,
        does_not=data.get("does_not") or [],
        examples=examples,
        checklists=data.get("checklists") or {},
    )


def list_role_slugs() -> list[str]:
    """Return the slugs of all available roles. README.md is excluded."""
    return [f.stem for f in sorted(_ROLES_DIR.iterdir()) if f.suffix == ".yaml"]


def load_role(slug: str) -> RolePrompt:
    """Load and parse a single role by slug.

    Raises:
        ValueError: If the role is not found in the embedded catalog.
    """
    if not _ROLE_SLUG_RE.fullmatch(slug):
        raise ValueError(f"Invalid role slug: {slug}")
    role_file = _ROLES_DIR / f"{slug}.yaml"
    if not role_file.exists():
        raise ValueError(f"Role not found: {slug}")
    with open(role_file) as f:
        data = yaml.safe_load(f)
    return _parse_role(data)


def load_role_catalog() -> dict[str, RolePrompt]:
    """Load all roles from the embedded catalog.

    Returns a dict keyed by slug.
    """
    catalog: dict[str, RolePrompt] = {}
    for slug in list_role_slugs():
        role = load_role(slug)
        catalog[role.slug] = role
    return catalog

File: lang/typescript/src/agentic.ts
import { readdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import * as yaml from "js-yaml";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const ROLES_DIR = join(__dirname, "../config/agentic/roles");

const ROLE_SLUG_RE = /^[a-z0-9][a-z0-9_-]*$/;

export interface RoleMindset {
  focus: string[];
  principles: string[];
}

export interface RoleEscalation {
  target: string;
  when: string;
}

export interface RoleExample {
  type: string;
  title: string;
  content: string;
}

/** Typed representation of a role definition from the agentic role catalog. */
export interface RolePrompt {
  slug: string;
  name: string;
  description: string;
  version: string;
  author?: string;
  status: string;
  category?: string;
  extends?: string;
  domains?: string[];
  tags?: string[];
  context?: string;
  scope: string[];
  mindset?: RoleMindset;
  responsibilities: string[];
  escalates_to: RoleEscalation[];
  does_not: string[];
  examples?: RoleExample[];
  checklists?: Record<string, string[]>;
}

/**
 * Returns the slugs of all available roles in the embedded catalog.
 * README.md is excluded automatically.
 */
export function listRoleSlugs(): string[] {
  const entries = readdirSync(ROLES_DIR);
  return entries
    .filter((f) => f.endsWith(".yaml"))
    .map((f) => f.replace(/\.yaml$/, ""))
    .sort();
}

/**
 * Loads and parses a single role by slug.
 * Returns undefined if the role is not found.
 */
export function loadRole(slug: string): RolePrompt | undefined {
  if (!ROLE_SLUG_RE.test(slug)) {
    return undefined;
  }
  try {
    const content = readFileSync(join(ROLES_DIR, `${slug}.yaml`), "utf-8");
    return yaml.load(content) as RolePrompt;
  } catch {
    return undefined;
  }
}

/**
 * Loads all roles from the embedded catalog.
 * Returns a Map keyed by slug.
 */
export function loadRoleCatalog(): Map<string, RolePrompt> {
  const catalog = new Map<string, RolePrompt>();
  for (const slug of listRoleSlugs()) {
    const role = loadRole(slug);
    if (role) {
      catalog.set(role.slug, role);
    }
  }
  return catalog;
}
